scripts/tftp-version.nse

```diff
@@ -0,0 +1,331 @@
+    if num ~= 1 then
+      return nil
+    end
+
+    -- The data remaining in the response must be from 0 to 512 bytes in length.
+    if #buf > 2 + 2 + 512 then
+      return nil
+    end
+
+    return ret
+  end
+
+  if opcode == OPCODE_ERROR then
+    -- The last byte in the packet must be zero to terminate the error message.
+    if buf:byte(#buf) ~= 0 then
+      return nil
+    end
+    ret.errcode = num
+
+    -- Extract the error message, if there is one.
+    if #buf > 2 + 2 + 1 then
+      ret.errmsg = ("z"):unpack(buf, 5)
+    end
+
+    return ret
+  end
+
+  -- Any other opcode, defined or otherwise, should not be coming back from the
+  -- service, so we treat it as an error.
+  return nil
+end
+
+action = function(host, port)
+  local output = stdnse.output_table()
+
+  -- Generate a random, unlikely filename in a format unlikely to be rejected,
+  -- specifically DOS 8.3 format.
+  local name = stdnse.generate_random_string(8, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_")
+  local extn = stdnse.generate_random_string(3, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
+  local path = name .. "." .. extn
+
+  -- Create and connect a socket.
+  local socket = nmap.new_socket("udp")
+  local status, err = socket:connect(host, port)
+  if not status then
+    socket:close()
+    output.ERROR = err
+    return output, output.ERROR
+  end
+
+  -- Remember the source port this socket used, for listening later.
+  local status, err, lport, _, _ = socket:get_info()
+  if not status then
+    socket:close()
+    output.ERROR = err
+    return output, output.ERROR
+  end
+
+  -- Generate a Read Request.
+  local req = (">Hzz"):pack(OPCODE_RRQ, path, "octet")
+
+  -- Send the Read Request.
+  socket:send(req)
+  socket:close()
+
+  -- Create a listening socket on the port from which we just sent.
+  local socket = nmap.new_socket("udp")
+  local status, err = socket:bind(nil, lport)
+  if not status then
+    socket:close()
+    output.ERROR = err
+    return output, output.ERROR
+  end
+
+  -- Listen for a response, but if nothing comes back we have to assume that
+  -- this is not a TFTP service and exit quietly.
+  --
+  -- We don't have to worry about other instance of this script running on other
+  -- ports of the same host confounding our results, because TFTP services
+  -- should respond back to the port matching the sending script.
+  --
+  -- Note that due to API limitations, we can't know if the response came from
+  -- the host we sent the request to, so we must assume it does.
+  local status, res = socket:receive()
+  socket:close()
+  if not status then
+    return nil
+  end
+
+  -- Parse the response.
+  local pkt = parse(res)
+  if not pkt then
+    return nil
+  end
+
+  -- Now we are convinced that the service speaks TFTP.
+  port.version.name = "tftp"
+
+  -- Populate the service information by referencing our list of software
+  -- responses.
+  identify_software(pkt, port)
```

Comment:
What about unidentified software? Have `identify_software` return the match info instead of setting it in the `port` object directly. Then do the `port.version` setting here in the action function; if there's no match, return output showing the message and code that were returned and requesting user to submit it to dev@nmap.org.